File: backend/axis/domain/country.py
"""Country: a deeply-modelled sovereign state.

Countries sit beneath the alliance/coalition layer (`Faction`). A country
belongs to exactly one faction (which may be neutral). Eight nested blocks
describe the country across government, military, nuclear posture,
demographics, diplomacy, energy/logistics, public opinion and geography.

All nested types are frozen, slotted dataclasses so the whole tree is hashable
and immutable after construction. Numeric ranges are validated in
`__post_init__` where appropriate.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum


def _check_unit_interval(name: str, value: float) -> None:
    if not 0.0 <= value <= 1.0:
        raise ValueError(f"{name} must be in [0, 1], got {value}")


def _check_signed_unit_interval(name: str, value: float) -> None:
    if not -1.0 <= value <= 1.0:
        raise ValueError(f"{name} must be in [-1, 1], got {value}")


# ---------------------------------------------------------------------------
# Government
# ---------------------------------------------------------------------------


class RegimeType(str, Enum):
    LIBERAL_DEMOCRACY = "liberal_democracy"
    ILLIBERAL_DEMOCRACY = "illiberal_democracy"
    HYBRID = "hybrid"
    AUTHORITARIAN = "authoritarian"
    MILITARY_JUNTA = "military_junta"


@dataclass(frozen=True, slots=True)
class CabinetMember:
    title: str
    name: str


@dataclass(frozen=True, slots=True)
class Government:
    regime_type: RegimeType
    head_of_state: str
    head_of_government: str
    cabinet: tuple[CabinetMember, ...] = field(default_factory=tuple)
    approval_rating: float = 0.5  # 0..1
    stability_index: float = 0.5  # 0..1
    last_election: str | None = None  # ISO date
    next_election: str | None = None  # ISO date

    def __post_init__(self) -> None:
        _check_unit_interval("Government.approval_rating", self.approval_rating)
        _check_unit_interval("Government.stability_index", self.stability_index)


# ---------------------------------------------------------------------------
# Military
# ---------------------------------------------------------------------------


class MilitaryPosture(str, Enum):
    DEFENSIVE = "defensive"
    DETERRENT = "deterrent"
    OFFENSIVE = "offensive"
    EXPEDITIONARY = "expeditionary"


class InventoryStatus(str, Enum):
    OPERATIONAL = "operational"
    LIMITED = "limited"
    RESERVE = "reserve"
    LEGACY = "legacy"


@dataclass(frozen=True, slots=True)
class InventoryLine:
    category: str  # e.g. "main_battle_tank", "fighter", "frigate"
    label: str  # e.g. "Leopard 2A6", "F-16C"
    count: int
    status: InventoryStatus = InventoryStatus.OPERATIONAL

    def __post_init__(self) -> None:
        if self.count < 0:
            raise ValueError(f"InventoryLine.count must be non-negative, got {self.count}")


@dataclass(frozen=True, slots=True)
class ServiceBranch:
    name: str  # e.g. "Lithuanian Land Force"
    personnel: int
    inventory: tuple[InventoryLine, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        if self.personnel < 0:
            raise ValueError("ServiceBranch.personnel must be non-negative")


@dataclass(frozen=True, slots=True)
class Military:
    active_personnel: int
    reserve_personnel: int
    paramilitary: int
    branches: tuple[ServiceBranch, ...]
    doctrine: str
    posture: MilitaryPosture
    alert_level: int  # 1..5 (5 = highest readiness)
    c2_nodes: tuple[str, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        for n, v in (
            ("active_personnel", self.active_personnel),
            ("reserve_personnel", self.reserve_personnel),
            ("paramilitary", self.paramilitary),
        ):
            if v < 0:
                raise ValueError(f"Military.{n} must be non-negative, got {v}")
        if not 1 <= self.alert_level <= 5:
            raise ValueError(f"Military.alert_level must be in [1, 5], got {self.alert_level}")


# ---------------------------------------------------------------------------
# Nuclear posture
# ---------------------------------------------------------------------------


class NuclearStatus(str, Enum):
    NWS = "nws"  # nuclear weapon state
    UMBRELLA_HOST = "umbrella_host"  # hosts allied warheads
    LATENT = "latent"  # technically capable, no declared programme
    NONE = "none"


@dataclass(frozen=True, slots=True)
class NuclearPosture:
    status: NuclearStatus
    warheads: int = 0
    delivery_systems: tuple[str, ...] = field(default_factory=tuple)
    declared_posture: str = ""
    nfu: bool | None = None  # no-first-use; None = undeclared / not applicable

    def __post_init__(self) -> None:
        if self.warheads < 0:
            raise ValueError("NuclearPosture.warheads must be non-negative")


# ---------------------------------------------------------------------------
# Demographics
# ---------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class Composition:
    label: str
    share: float  # 0..1

    def __post_init__(self) -> None:
        _check_unit_interval(f"Composition[{self.label}].share", self.share)


@dataclass(frozen=True, slots=True)
class Demographics:
    population: int
    median_age: float
    urbanisation: float  # 0..1
    ethnic_groups: tuple[Composition, ...] = field(default_factory=tuple)
    languages: tuple[Composition, ...] = field(default_factory=tuple)
    religions: tuple[Composition, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        if self.population < 0:
            raise ValueError("Demographics.population must be non-negative")
        if self.median_age < 0:
            raise ValueError("Demographics.median_age must be non-negative")
        _check_unit_interval("Demographics.urbanisation", self.urbanisation)


# ---------------------------------------------------------------------------
# Diplomacy
# ---------------------------------------------------------------------------


class RelationStatus(str, Enum):
    ALLIED = "allied"
    FRIENDLY = "friendly"
    NEUTRAL = "neutral"
    STRAINED = "strained"
    HOSTILE = "hostile"
    AT_WAR = "at_war"


@dataclass(frozen=True, slots=True)
class Treaty:
    name: str
    kind: str  # e.g. "collective_defence", "arms_control", "trade"
    parties: tuple[str, ...] = field(default_factory=tuple)
    in_force: bool = True


@dataclass(frozen=True, slots=True)
class BilateralRelation:
    other_country_id: str  # may reference a country not yet seeded; FE handles gracefully
    status: RelationStatus
    score: float  # -1..+1, signed quality of the relationship

    def __post_init__(self) -> None:
        _check_signed_unit_interval("BilateralRelation.score", self.score)


@dataclass(frozen=True, slots=True)
class Diplomacy:
    alliance_memberships: tuple[str, ...] = field(default_factory=tuple)
    treaties: tuple[Treaty, ...] = field(default_factory=tuple)
    relations: tuple[BilateralRelation, ...] = field(default_factory=tuple)


# ---------------------------------------------------------------------------
# Energy / logistics
# ---------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class EnergyLogistics:
    oil_dependence: float  # 0..1, share imported
    gas_dependence: float  # 0..1, share imported
    top_gas_supplier: str
    pipelines: tuple[str, ...] = field(default_factory=tuple)
    key_ports: tuple[str, ...] = field(default_factory=tuple)
    rail_gauge_mm: int = 1435  # 1435 standard, 1520 Russian/CIS broad
    strategic_reserves_days: int = 0

    def __post_init__(self) -> None:
        _check_unit_interval("EnergyLogistics.oil_dependence", self.oil_dependence)
        _check_unit_interval("EnergyLogistics.gas_dependence", self.gas_dependence)
        if self.rail_gauge_mm <= 0:
            raise ValueError("EnergyLogistics.rail_gauge_mm must be positive")
        if self.strategic_reserves_days < 0:
            raise ValueError("EnergyLogistics.strategic_reserves_days must be non-negative")


# ---------------------------------------------------------------------------
# Public opinion
# ---------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class PublicOpinion:
    war_support: float  # 0..1, share of public supportive of armed action
    institutional_trust: float  # 0..1
    censorship_index: float  # 0..1, higher = more state media control
    protest_intensity: float  # 0..1
    top_outlets: tuple[str, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        _check_unit_interval("PublicOpinion.war_support", self.war_support)
        _check_unit_interval("PublicOpinion.institutional_trust", self.institutional_trust)
        _check_unit_interval("PublicOpinion.censorship_index", self.censorship_index)
        _check_unit_interval("PublicOpinion.protest_intensity", self.protest_intensity)


# ---------------------------------------------------------------------------
# Geography
# ---------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class Border:
    other: str  # country id where known, otherwise display name
    length_km: int

    def __post_init__(self) -> None:
        if self.length_km < 0:
            raise ValueError("Border.length_km must be non-negative")


@dataclass(frozen=True, slots=True)
class KeyBase:
    name: str
    kind: str  # e.g. "air_base", "naval_base", "command_node", "training_ground"
    lon: float
    lat: float
    owner_country_id: str

    def __post_init__(self) -> None:
        if not -180.0 <= self.lon <= 180.0:
            raise ValueError(f"KeyBase.lon out of range: {self.lon}")
        if not -90.0 <= self.lat <= 90.0:
            raise ValueError(f"KeyBase.lat out of range: {self.lat}")


@dataclass(frozen=True, slots=True)
class Geography:
    area_km2: int
    land_borders: tuple[Border, ...] = field(default_factory=tuple)
    key_bases: tuple[KeyBase, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        if self.area_km2 < 0:
            raise ValueError("Geography.area_km2 must be non-negative")


# ---------------------------------------------------------------------------
# Country (top-level)
# ---------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class Country:
    """A sovereign state with deep nested attributes.

    `faction_id` references the alliance/coalition layer (NATO, CSTO,
    neutral). All structural numeric constraints are enforced by the nested
    dataclasses; this top-level class only checks identifier presence.
    """

    id: str
    iso_a2: str  # ISO 3166-1 alpha-2
    iso_a3: str  # ISO 3166-1 alpha-3
    name: str
    official_name: str
    faction_id: str
    flag_emoji: str
    capital_city_id: str | None
    government: Government
    military: Military
    nuclear: NuclearPosture
    demographics: Demographics
    diplomacy: Diplomacy
    energy: EnergyLogistics
    public_opinion: PublicOpinion
    geography: Geography
    available_actions: tuple[str, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        if not self.id:
            raise ValueError("Country.id must be non-empty")
        if len(self.iso_a2) != 2:
            raise ValueError(f"Country.iso_a2 must be 2 chars, got {self.iso_a2!r}")
        if len(self.iso_a3) != 3:
            raise ValueError(f"Country.iso_a3 must be 3 chars, got {self.iso_a3!r}")

File: frontend/src/types/country.ts
import { z } from "zod";

export const regimeTypeSchema = z.enum([
  "liberal_democracy",
  "illiberal_democracy",
  "hybrid",
  "authoritarian",
  "military_junta",
]);
export type RegimeType = z.infer<typeof regimeTypeSchema>;

export const militaryPostureSchema = z.enum([
  "defensive",
  "deterrent",
  "offensive",
  "expeditionary",
]);
export type MilitaryPosture = z.infer<typeof militaryPostureSchema>;

export const inventoryStatusSchema = z.enum([
  "operational",
  "limited",
  "reserve",
  "legacy",
]);
export type InventoryStatus = z.infer<typeof inventoryStatusSchema>;

export const nuclearStatusSchema = z.enum([
  "nws",
  "umbrella_host",
  "latent",
  "none",
]);
export type NuclearStatus = z.infer<typeof nuclearStatusSchema>;

export const relationStatusSchema = z.enum([
  "allied",
  "friendly",
  "neutral",
  "strained",
  "hostile",
  "at_war",
]);
export type RelationStatus = z.infer<typeof relationStatusSchema>;

const compositionSchema = z.object({
  label: z.string(),
  share: z.number().min(0).max(1),
});
export type Composition = z.infer<typeof compositionSchema>;

const cabinetMemberSchema = z.object({
  title: z.string(),
  name: z.string(),
});
export type CabinetMember = z.infer<typeof cabinetMemberSchema>;

const governmentSchema = z.object({
  regime_type: regimeTypeSchema,
  head_of_state: z.string(),
  head_of_government: z.string(),
  cabinet: z.array(cabinetMemberSchema).default([]),
  approval_rating: z.number().min(0).max(1),
  stability_index: z.number().min(0).max(1),
  last_election: z.string().nullable().optional(),
  next_election: z.string().nullable().optional(),
});
export type Government = z.infer<typeof governmentSchema>;

const inventoryLineSchema = z.object({
  category: z.string(),
  label: z.string(),
  count: z.number().int().nonnegative(),
  status: inventoryStatusSchema,
});
export type InventoryLine = z.infer<typeof inventoryLineSchema>;

const serviceBranchSchema = z.object({
  name: z.string(),
  personnel: z.number().int().nonnegative(),
  inventory: z.array(inventoryLineSchema).default([]),
});
export type ServiceBranch = z.infer<typeof serviceBranchSchema>;

const militarySchema = z.object({
  active_personnel: z.number().int().nonnegative(),
  reserve_personnel: z.number().int().nonnegative(),
  paramilitary: z.number().int().nonnegative(),
  branches: z.array(serviceBranchSchema).default([]),
  doctrine: z.string(),
  posture: militaryPostureSchema,
  alert_level: z.number().int().min(1).max(5),
  c2_nodes: z.array(z.string()).default([]),
});
export type Military = z.infer<typeof militarySchema>;

const nuclearSchema = z.object({
  status: nuclearStatusSchema,
  warheads: z.number().int().nonnegative(),
  delivery_systems: z.array(z.string()).default([]),
  declared_posture: z.string(),
  nfu: z.boolean().nullable(),
});
export type NuclearPosture = z.infer<typeof nuclearSchema>;

const demographicsSchema = z.object({
  population: z.number().int().nonnegative(),
  median_age: z.number().nonnegative(),
  urbanisation: z.number().min(0).max(1),
  ethnic_groups: z.array(compositionSchema).default([]),
  languages: z.array(compositionSchema).default([]),
  religions: z.array(compositionSchema).default([]),
});
export type Demographics = z.infer<typeof demographicsSchema>;

const treatySchema = z.object({
  name: z.string(),
  kind: z.string(),
  parties: z.array(z.string()).default([]),
  in_force: z.boolean(),
});
export type Treaty = z.infer<typeof treatySchema>;

const bilateralRelationSchema = z.object({
  other_country_id: z.string(),
  status: relationStatusSchema,
  score: z.number().min(-1).max(1),
});
export type BilateralRelation = z.infer<typeof bilateralRelationSchema>;

const diplomacySchema = z.object({
  alliance_memberships: z.array(z.string()).default([]),
  treaties: z.array(treatySchema).default([]),
  relations: z.array(bilateralRelationSchema).default([]),
});
export type Diplomacy = z.infer<typeof diplomacySchema>;

const energySchema = z.object({
  oil_dependence: z.number().min(0).max(1),
  gas_dependence: z.number().min(0).max(1),
  top_gas_supplier: z.string(),
  pipelines: z.array(z.string()).default([]),
  key_ports: z.array(z.string()).default([]),
  rail_gauge_mm: z.number().int().positive(),
  strategic_reserves_days: z.number().int().nonnegative(),
});
export type EnergyLogistics = z.infer<typeof energySchema>;

const publicOpinionSchema = z.object({
  war_support: z.number().min(0).max(1),
  institutional_trust: z.number().min(0).max(1),
  censorship_index: z.number().min(0).max(1),
  protest_intensity: z.number().min(0).max(1),
  top_outlets: z.array(z.string()).default([]),
});
export type PublicOpinion = z.infer<typeof publicOpinionSchema>;

const borderSchema = z.object({
  other: z.string(),
  length_km: z.number().int().nonnegative(),
});
export type Border = z.infer<typeof borderSchema>;

const keyBaseSchema = z.object({
  name: z.string(),
  kind: z.string(),
  lon: z.number(),
  lat: z.number(),
  owner_country_id: z.string(),
});
export type KeyBase = z.infer<typeof keyBaseSchema>;

const geographySchema = z.object({
  area_km2: z.number().int().nonnegative(),
  land_borders: z.array(borderSchema).default([]),
  key_bases: z.array(keyBaseSchema).default([]),
});
export type Geography = z.infer<typeof geographySchema>;

export const countrySchema = z.object({
  id: z.string(),
  iso_a2: z.string().length(2),
  iso_a3: z.string().length(3),
  name: z.string(),
  official_name: z.string(),
  faction_id: z.string(),
  flag_emoji: z.string(),
  capital_city_id: z.string().nullable().optional(),
  government: governmentSchema,
  military: militarySchema,
  nuclear: nuclearSchema,
  demographics: demographicsSchema,
  diplomacy: diplomacySchema,
  energy: energySchema,
  public_opinion: publicOpinionSchema,
  geography: geographySchema,
  available_actions: z.array(z.string()).default([]),
});
export type Country = z.infer<typeof countrySchema>;

/** Metrics that can shade the territory choropleth. */
export type ChoroplethMetric =
  | "war_support"
  | "alert_level"
  | "stability_index"
  | "approval_rating"
  | "protest_intensity"
  | "censorship_index"
  | "institutional_trust";

export const CHOROPLETH_METRICS: { id: ChoroplethMetric; label: string }[] = [
  { id: "war_support", label: "War support" },
  { id: "alert_level", label: "Alert level" },
  { id: "stability_index", label: "Stability" },
  { id: "approval_rating", label: "Govt approval" },
  { id: "protest_intensity", label: "Protest intensity" },
  { id: "censorship_index", label: "Censorship" },
  { id: "institutional_trust", label: "Institutional trust" },
];

/** Normalise a metric to 0..1 for shading. Higher = "better" for the holder. */
export function metricNormalised(country: Country, metric: ChoroplethMetric): number {
  switch (metric) {
    case "war_support":
      return country.public_opinion.war_support;
    case "alert_level":
      // 1..5 -> 0..1
      return (country.military.alert_level - 1) / 4;
    case "stability_index":
      return country.government.stability_index;
    case "approval_rating":
      return country.government.approval_rating;
    case "protest_intensity":
      return country.public_opinion.protest_intensity;
    case "censorship_index":
      return country.public_opinion.censorship_index;
    case "institutional_trust":
      return country.public_opinion.institutional_trust;
  }
}

/** Display label for a raw metric value (preserves units). */
export function metricDisplay(country: Country, metric: ChoroplethMetric): string {
  switch (metric) {
    case "alert_level":
      return `${country.military.alert_level} / 5`;
    default:
      return `${Math.round(metricNormalised(country, metric) * 100)}%`;
  }
}
